dask/dataframe/core.py

```diff
@@ -2788,36 +2793,39 @@ def map_partitions(func, *args, **kwargs):
         # If `meta` is not a pandas object, the concatenated results will be a
         # different type
         meta = _concat([meta])
-
-    if isinstance(meta, pd.DataFrame):
-        columns = meta.columns
-    elif isinstance(meta, (pd.Series, pd.Index)):
-        columns = meta.name
-    else:
-        columns = None
+    meta = make_meta(meta)
 
     dfs = [df for df in args if isinstance(df, _Frame)]
     dsk = {}
     for i in range(dfs[0].npartitions):
         values = [(arg._name, i if isinstance(arg, _Frame) else 0)
                   if isinstance(arg, (_Frame, Scalar)) else arg for arg in args]
-        values = (apply, func, (tuple, values), kwargs)
-        if columns is not None:
-            values = (_rename, columns, values)
-        dsk[(name, i)] = values
+        dsk[(name, i)] = (apply_and_enforce, func, values, kwargs, meta)
```

Comment:
Note that this change now serializes the empty `meta` object as part of the graph. This is so it can be used when the result is empty, and to pass along the column name(s). This could be relaxed back to the previous behavior of only passing along names, and have the empty partitions just be of `object` dtype.

As an aside, I'm not sure why the renaming behavior exists in `map_partitions`, but it does, and there's tests for it.